<a href="https://colab.research.google.com/github/jieyingc/fast-failure-detection/blob/main/explore_dataset_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from datasets import load_dataset

dataset = load_dataset("SWE-bench/SWE-smith-trajectories")

print(dataset)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md: 0.00B [00:00, ?B/s]

data/tool-00000-of-00008.parquet:   0%|          | 0.00/249M [00:00<?, ?B/s]

data/tool-00001-of-00008.parquet:   0%|          | 0.00/116M [00:00<?, ?B/s]

data/tool-00002-of-00008.parquet:   0%|          | 0.00/99.6M [00:00<?, ?B/s]

data/tool-00003-of-00008.parquet:   0%|          | 0.00/138M [00:00<?, ?B/s]

data/tool-00004-of-00008.parquet:   0%|          | 0.00/77.8M [00:00<?, ?B/s]

data/tool-00005-of-00008.parquet:   0%|          | 0.00/123M [00:00<?, ?B/s]

data/tool-00006-of-00008.parquet:   0%|          | 0.00/153M [00:00<?, ?B/s]

data/tool-00007-of-00008.parquet:   0%|          | 0.00/122M [00:00<?, ?B/s]

data/xml-00000-of-00008.parquet:   0%|          | 0.00/106M [00:00<?, ?B/s]

data/xml-00001-of-00008.parquet:   0%|          | 0.00/121M [00:00<?, ?B/s]

data/xml-00002-of-00008.parquet:   0%|          | 0.00/115M [00:00<?, ?B/s]

data/xml-00003-of-00008.parquet:   0%|          | 0.00/144M [00:00<?, ?B/s]

data/xml-00004-of-00008.parquet:   0%|          | 0.00/126M [00:00<?, ?B/s]

data/xml-00005-of-00008.parquet:   0%|          | 0.00/244M [00:00<?, ?B/s]

data/xml-00006-of-00008.parquet:   0%|          | 0.00/113M [00:00<?, ?B/s]

data/xml-00007-of-00008.parquet:   0%|          | 0.00/107M [00:00<?, ?B/s]

data/ticks-00000-of-00008.parquet:   0%|          | 0.00/114M [00:00<?, ?B/s]

data/ticks-00001-of-00008.parquet:   0%|          | 0.00/94.6M [00:00<?, ?B/s]

data/ticks-00002-of-00008.parquet:   0%|          | 0.00/112M [00:00<?, ?B/s]

data/ticks-00003-of-00008.parquet:   0%|          | 0.00/143M [00:00<?, ?B/s]

data/ticks-00004-of-00008.parquet:   0%|          | 0.00/110M [00:00<?, ?B/s]

data/ticks-00005-of-00008.parquet:   0%|          | 0.00/95.7M [00:00<?, ?B/s]

data/ticks-00006-of-00008.parquet:   0%|          | 0.00/120M [00:00<?, ?B/s]

data/ticks-00007-of-00008.parquet:   0%|          | 0.00/229M [00:00<?, ?B/s]

Generating tool split:   0%|          | 0/24100 [00:00<?, ? examples/s]

Generating xml split:   0%|          | 0/26076 [00:00<?, ? examples/s]

Generating ticks split:   0%|          | 0/25826 [00:00<?, ? examples/s]

DatasetDict({
    tool: Dataset({
        features: ['messages', 'instance_id', 'resolved', 'model', 'traj_id', 'patch'],
        num_rows: 24100
    })
    xml: Dataset({
        features: ['messages', 'instance_id', 'resolved', 'model', 'traj_id', 'patch'],
        num_rows: 26076
    })
    ticks: Dataset({
        features: ['messages', 'instance_id', 'resolved', 'model', 'traj_id', 'patch'],
        num_rows: 25826
    })
})


In [2]:
sample = dataset["tool"][0]

for k, v in sample.items():
    print("FIELD:", k, type(k))
    print(v)
    print("=" * 80)

FIELD: messages <class 'str'>
[{"role": "system", "content": "You are a helpful assistant that can interact with a computer to solve tasks.", "agent": "main", "message_type": "system_prompt"}, {"role": "user", "content": [{"type": "text", "text": "<uploaded_files>\n/testbed\n</uploaded_files>\nI've uploaded a python code repository in the directory /testbed. Consider the following PR description:\n\n<pr_description>\n# MoneyWidget decompress method breaks form validation with disabled fields\n\nI've discovered an issue with the `MoneyWidget` class in the forms/widgets.py file. When a form field using this widget is disabled, validation fails unexpectedly.\n\n## Steps to reproduce\n\n1. Create a model with a MoneyField:\n```python\nclass ModelWithVanillaMoneyField(models.Model):\n    money = MoneyField(max_digits=10, decimal_places=2, default_currency='USD')\n```\n\n2. Create a form with a disabled money field:\n```python\nclass DisabledFieldForm(forms.ModelForm):\n    class Meta:\n    

In [3]:
from collections import Counter

print(Counter(dataset["tool"]["resolved"]))
print(Counter(dataset["xml"]["resolved"]))
print(Counter(dataset["ticks"]["resolved"]))

Counter({False: 14673, True: 9427})
Counter({False: 14573, True: 11503})
Counter({False: 14396, True: 11430})


In [4]:
for split in dataset.keys():
    print(f"\n=== {split} ===")
    print(Counter(dataset[split]["model"]).most_common(10))


=== tool ===
[('claude-3-7-sonnet-20250219', 17715), ('claude-3-5-sonnet-20241022', 5751), ('gpt-4o-2024-08-06', 634)]

=== xml ===
[('claude-3-7-sonnet-20250219', 18457), ('claude-3-5-sonnet-20241022', 7091), ('gpt-4o-2024-08-06', 528)]

=== ticks ===
[('claude-3-7-sonnet-20250219', 18457), ('claude-3-5-sonnet-20241022', 6894), ('gpt-4o-2024-08-06', 475)]


In [5]:
tool_ids = set(dataset["tool"]["instance_id"])
xml_ids = set(dataset["xml"]["instance_id"])
ticks_ids = set(dataset["ticks"]["instance_id"])

print("tool ∩ xml:", len(tool_ids & xml_ids))
print("tool ∩ ticks:", len(tool_ids & ticks_ids))
print("xml ∩ ticks:", len(xml_ids & ticks_ids))

tool ∩ xml: 10637
tool ∩ ticks: 10636
xml ∩ ticks: 13500


- dataset 由 tool/xml/ticks 三个子集构成，不是标准 train/val/test
- 三个子集之间 instance_id 大量重叠，不能直接拿来当 train/val/test

### **Check null and empty**
- patch 字段有较高比例为空，但 empty patch 并不意味着 agent 没有完成修复
- resolved=True 的 empty-patch 样本仍然包含完整的修复与提交轨迹

In [6]:
for col in dataset["tool"].column_names:
    null_count = sum(row[col] is None for row in dataset["tool"])
    print(f"{col}: {null_count} nulls")

messages: 0 nulls
instance_id: 0 nulls
resolved: 0 nulls
model: 0 nulls
traj_id: 0 nulls
patch: 0 nulls


In [ ]:
def check_empty_strings(hf_dataset):
    for col in hf_dataset.column_names:
        empty_count = 0
        for row in hf_dataset:
            value = row[col]
            if isinstance(value, str) and value.strip() == "":
                empty_count += 1
        print(f"{col}: {empty_count} empty strings")

print(check_empty_strings(dataset["tool"]))
print(check_empty_strings(dataset["xml"]))
print(check_empty_strings(dataset["ticks"]))

messages: 0 empty strings
instance_id: 0 empty strings
resolved: 0 empty strings
model: 0 empty strings
traj_id: 0 empty strings
patch: 5944 empty strings
None
messages: 0 empty strings
instance_id: 0 empty strings
resolved: 0 empty strings
model: 0 empty strings
traj_id: 0 empty strings
patch: 7352 empty strings
None
messages: 0 empty strings
instance_id: 0 empty strings
resolved: 0 empty strings
model: 0 empty strings
traj_id: 0 empty strings
patch: 7279 empty strings
None


In [7]:
empty_patch_labels = Counter()
nonempty_patch_labels = Counter()

for row in dataset["tool"]:
    if row["patch"].strip() == "":
        empty_patch_labels[row["resolved"]] += 1
    else:
        nonempty_patch_labels[row["resolved"]] += 1

print("Empty patch labels:", empty_patch_labels)
print("Non-empty patch labels:", nonempty_patch_labels)

Empty patch labels: Counter({False: 3662, True: 2282})
Non-empty patch labels: Counter({False: 11011, True: 7145})


In [13]:
import json

rows = [row for row in dataset["tool"] if row["patch"].strip() == "" and row["resolved"]]

for row in rows[:3]:
    msgs = json.loads(row["messages"])
    print("INSTANCE:", row["instance_id"])
    print("MODEL:", row["model"])
    print("NUM_MESSAGES:", len(msgs))
    print("LAST 3 MESSAGES:")
    for m in msgs[-3:]:
        print(m)
    print("=" * 80)

INSTANCE: django-money__django-money.835c1ab8.func_pm_ctrl_shuffle__viqnyl9u
MODEL: claude-3-7-sonnet-20250219
NUM_MESSAGES: 49
LAST 3 MESSAGES:
{'role': 'assistant', 'content': 'Let me remove the reproduction script and submit again:', 'thought': 'Let me remove the reproduction script and submit again:', 'action': 'rm /testbed/reproduce_issue.py', 'agent': 'main', 'tool_calls': [{'index': 1, 'function': {'arguments': '{"command": "rm /testbed/reproduce_issue.py"}', 'name': 'bash'}, 'id': 'toolu_01RX7JN3tbDwDNYoAFzeEDaL', 'type': 'function'}], 'message_type': 'action'}
{'role': 'tool', 'content': [{'type': 'text', 'text': 'Your command ran successfully and did not produce any output.'}], 'agent': 'main', 'message_type': 'observation', 'tool_call_ids': ['toolu_01RX7JN3tbDwDNYoAFzeEDaL'], 'cache_control': {'type': 'ephemeral'}}
{'role': 'assistant', 'content': '', 'thought': '', 'action': 'submit', 'agent': 'main', 'tool_calls': [{'index': 0, 'function': {'arguments': '{}', 'name': 'subm

**split by instance_id not by row**:
- tool 子集中有 10,637 个 unique instances 和 24,100 条 trajectories，说明一个 instance 可对应多次 agent 尝试

In [8]:
tool_instance_counts = Counter(dataset["tool"]["instance_id"])
dup_tool = sum(c > 1 for c in tool_instance_counts.values())
max_count = max(tool_instance_counts.values())

print("unique instance_ids:", len(tool_instance_counts))
print("duplicated instance_ids:", dup_tool)
print("max trajectories per instance:", max_count)

unique instance_ids: 10637
duplicated instance_ids: 4675
max trajectories per instance: 21


In [ ]:
import json
import re


def extract_features(row):
    patch = row["patch"] if row["patch"] is not None else ""
    patch_stripped = patch.strip()

    # Parse messages
    try:
        msgs = json.loads(row["messages"])
    except Exception:
        msgs = []

    # -------------------------
    # 1) Patch-level features
    # -------------------------
    patch_is_empty = int(patch_stripped == "")
    patch_char_len = len(patch)

    # Count added / removed lines, excluding diff headers like +++ / ---
    num_lines_added = 0
    num_lines_removed = 0

    for line in patch.splitlines():
        if line.startswith("+++ ") or line.startswith("--- "):
            continue
        if line.startswith("+"):
            num_lines_added += 1
        elif line.startswith("-"):
            num_lines_removed += 1

    # Number of files touched
    num_files_touched = patch.count("diff --git")

    # Whether test files were modified
    touched_files = re.findall(r"diff --git a/(.*?) b/(.*?)\n", patch)
    test_file_modified = 0
    for a_path, b_path in touched_files:
        path = b_path.lower()
        if (
            "test" in path
            or "tests/" in path
            or path.endswith("_test.py")
            or "/test_" in path
        ):
            test_file_modified = 1
            break

    # Very simple repeated-block heuristic:
    # if the same added/removed code line repeats many times
    code_lines = []
    for line in patch.splitlines():
        if line.startswith("+++ ") or line.startswith("--- ") or line.startswith("@@"):
            continue
        if line.startswith("+") or line.startswith("-"):
            content = line[1:].strip()
            if content:
                code_lines.append(content)

    contain_repeated_blocks = 0
    if code_lines:
        line_freq = {}
        for x in code_lines:
            line_freq[x] = line_freq.get(x, 0) + 1
        if max(line_freq.values()) >= 3:
            contain_repeated_blocks = 1

    # -------------------------
    # 2) Message / trajectory features
    # -------------------------
    num_messages = len(msgs)
    messages_char_len = len(row["messages"])

    assistant_num_messages = 0
    user_num_messages = 0
    tool_num_messages = 0
    system_num_messages = 0

    num_action_messages = 0
    num_observation_messages = 0

    has_submit = 0
    num_submit_calls = 0

    has_bash = 0
    num_bash_calls = 0

    has_str_replace_editor = 0
    num_str_replace_editor_calls = 0

    num_edit_calls = 0
    num_error_messages = 0

    # Keywords for simple error detection
    error_keywords = [
        "error",
        "exception",
        "traceback",
        "failed",
        "failure",
        "syntaxerror",
        "typeerror",
        "valueerror",
        "assertionerror",
    ]

    # tool/function names counted as editing-related
    edit_tool_names = {
        "str_replace_editor",
        "edit_file",
        "create_file",
        "insert",
        "replace",
        "write_file",
    }

    for msg in msgs:
        role = msg.get("role", "")
        message_type = msg.get("message_type", "")

        # Role counts
        if role == "assistant":
            assistant_num_messages += 1
        elif role == "user":
            user_num_messages += 1
        elif role == "tool":
            tool_num_messages += 1
        elif role == "system":
            system_num_messages += 1

        # Message-type counts
        if message_type == "action":
            num_action_messages += 1
        elif message_type == "observation":
            num_observation_messages += 1

        # Check action field
        action_text = str(msg.get("action", "")).lower()

        if action_text == "submit" or "submit" in action_text:
            has_submit = 1

        # Collect all text for simple error matching
        text_blobs = []

        content = msg.get("content", "")
        if isinstance(content, str):
            text_blobs.append(content.lower())
        elif isinstance(content, list):
            for item in content:
                if isinstance(item, dict):
                    t = item.get("text", "")
                    if isinstance(t, str):
                        text_blobs.append(t.lower())

        thought = msg.get("thought", "")
        if isinstance(thought, str):
            text_blobs.append(thought.lower())

        if isinstance(action_text, str):
            text_blobs.append(action_text)

        # tool_calls parsing
        tool_calls = msg.get("tool_calls", [])
        if isinstance(tool_calls, list):
            for tc in tool_calls:
                fn_name = (
                    tc.get("function", {}).get("name", "")
                    if isinstance(tc, dict)
                    else ""
                )
                fn_name = str(fn_name).lower()

                if fn_name == "submit":
                    has_submit = 1
                    num_submit_calls += 1

                if fn_name == "bash":
                    has_bash = 1
                    num_bash_calls += 1

                if fn_name == "str_replace_editor":
                    has_str_replace_editor = 1
                    num_str_replace_editor_calls += 1

                if fn_name in edit_tool_names or fn_name == "str_replace_editor":
                    num_edit_calls += 1

        # Also infer from action text
        if "bash" in action_text:
            has_bash = 1

        if "str_replace_editor" in action_text:
            has_str_replace_editor = 1

        # Simple error keyword counting
        joined_text = " ".join(text_blobs)
        for kw in error_keywords:
            num_error_messages += joined_text.count(kw)

    features = {
        # IDs / metadata
        "instance_id": row["instance_id"],
        "traj_id": row["traj_id"],
        "model": row["model"],

        # Label
        "resolved": int(bool(row["resolved"])),

        # Patch features
        "patch_is_empty": patch_is_empty,
        "patch_char_len": patch_char_len,
        "num_lines_added": num_lines_added,
        "num_lines_removed": num_lines_removed,
        "num_files_touched": num_files_touched,
        "test_file_modified": test_file_modified,
        "contain_repeated_blocks": contain_repeated_blocks,

        # Message / trajectory features
        "num_messages": num_messages,
        "messages_char_len": messages_char_len,
        "assistant_num_messages": assistant_num_messages,
        "user_num_messages": user_num_messages,
        "tool_num_messages": tool_num_messages,
        "system_num_messages": system_num_messages,
        "num_action_messages": num_action_messages,
        "num_observation_messages": num_observation_messages,
        "has_submit": has_submit,
        "num_submit_calls": num_submit_calls,
        "has_bash": has_bash,
        "num_bash_calls": num_bash_calls,
        "has_str_replace_editor": has_str_replace_editor,
        "num_str_replace_editor_calls": num_str_replace_editor_calls,
        "num_edit_calls": num_edit_calls,
        "num_error_messages": num_error_messages,
    }

    return features